RFCs/FS-1135-random-functions-for-collections.md

```diff
@@ -0,0 +1,175 @@
+val shuffleRand: Random -> CollectionType<'T> -> CollectionType<'T>
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.shuffle // [ "Charlie"; "Dave"; "Alice"; "Bob" ]
+```
+
+
+### Choice
+
+Choice function should returned a random element from the collection.
+
+Two functions should be added to each module.
+
+```fsharp
+val choice: CollectionType<'T> -> 'T
+val choiceRand: Random -> CollectionType<'T> -> 'T
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
+
+[InvalidOperationException](https://learn.microsoft.com/en-us/dotnet/api/system.invalidoperationexception) should be raised if collection is empty
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.choice // "Charlie"
+```
+
+
+### Choices
+
+Choices should select N elements from input collection in random order, once element is taken it can be selected again.
+
+Two functions should be added to each module.
+
+```fsharp
+val choices: int -> CollectionType<'T> -> CollectionType<'T>
+val choicesRand: Random -> int -> CollectionType<'T> -> CollectionType<'T>
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
+
+[ArgumentOutOfRangeException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception) should be raised if N is negative
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.choices 3 // ["Bob", "Dave", "Bob"]
+```
+
+### Sample
+
+Sample should select N elements from input collection in random order, once element is taken it won't be selected again. N can't be greater than collection length
+
+Two functions should be added to each module.
+
+```fsharp
+val sample: int -> CollectionType<'T> -> CollectionType<'T>
+val sampleRand: Random -> int -> CollectionType<'T> -> CollectionType<'T>
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
+
+[ArgumentOutOfRangeException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception) should be raised if N is greater than collection length or negative
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.sample 3 // ["Charlie", "Dave", "Alice"]
+```
+
+# Drawbacks
+
+System.Random interface has added some new methods in [.NET 8](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8#methods-for-working-with-randomness), where naming is a bit different. More new methods can eventually be added in future .NET versions.
```

Comment:
Ehhh, I think it's pretty unlikely that we could add these in. Maybe in the far, far future we can expect .NET 8 as a base dependency for FSharp.Core, but that's so far away that it's not even worth considering.

Reply:
Yeah, +1 we should use/target only what's available in ns2.0 at this point.

Reply:
I've marked this unresolved question as resolved in the end of the RFC. The drawback itself I think should stay, since users will be choosing - to use .NET shuffling and sampling or F# shuffling and sampling

Reply:
I think what you mean to say here is that we're adding on top of an existing surface area of methods in .NET. I don't think you propose here to use, or refer to these new methods? (they are not part of NetStandard).

However, that said, I'm not opposed mentioning it, but maybe more along the lines of:

```suggestion
Users may be tempted to use some of the recently added method of `System.Random` that also apply to collections instead of the ones we add in FSharp.Core. It may also be confusing to some, especially since the naming over there is slightly different. See [.NET 8 What's New](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8#methods-for-working-with-randomness).
```

It's certainly not an unresolved question ;).

Reply:
This has since been resolved